File: frontend/src/api/fileTypes.ts
import type { FileItem, PreviewType } from './files'

const STRUCTURE_FALLBACK_EXTENSIONS = new Set([
  '.xyz',
  '.pdb',
  '.mol',
  '.sdf',
  '.cif',
  '.traj',
  '.extxyz',
  '.vasp',
  '.xsf',
  '.xsd',
  '.xtd',
  '.arc',
  '.cube',
  '.gen',
  '.db',
  '.lammps',
  '.dump',
  '.gjf',
  '.com',
  '.fdf',
  '.pwi'
])

const STRUCTURE_FALLBACK_NAMES = new Set(['vasp.xml', 'vasprun.xml'])
const VASP_STRUCTURE_NAME_RE = /(^|[^A-Z0-9])(POSCAR|CONTCAR|XDATCAR|OUTCAR)(?:[0-9]+)?($|[^A-Z0-9])/

export function pathBaseName(path: string) {
  return path.split(/[\\/]/).pop() ?? path
}

export function extensionFromName(name: string) {
  const index = name.lastIndexOf('.')
  return index >= 0 ? name.slice(index).toLowerCase() : ''
}

export function isForcedStructureName(name: string) {
  return VASP_STRUCTURE_NAME_RE.test(name.toUpperCase())
}

export function isStructurePreviewItem(item: Pick<FileItem, 'name' | 'preview_type'>) {
  return item.preview_type === 'structure' || isForcedStructureName(item.name)
}

export function isStructurePreviewPath(path: string, previewType?: PreviewType | null) {
  if (previewType === 'structure') return true
  const name = pathBaseName(path).toLowerCase()
  return (
    isForcedStructureName(name) ||
    STRUCTURE_FALLBACK_NAMES.has(name) ||
    STRUCTURE_FALLBACK_EXTENSIONS.has(extensionFromName(name))
  )
}

File: backend/app/services/file_types.py
from __future__ import annotations

import re
from pathlib import Path
from typing import Optional


STRUCTURE_EXTENSIONS = {
    ".xyz": "xyz",
    ".pdb": "pdb",
    ".mol": "mol",
    ".sdf": "sdf",
    ".cif": "cif",
    ".traj": "traj",
    ".extxyz": "extxyz",
    ".vasp": "vasp",
    ".xsf": "xsf",
    ".xsd": "xsd",
    ".xtd": "xtd",
    ".arc": "dmol-arc",
    ".cube": "cube",
    ".gen": "gen",
    ".db": "db",
    ".lammps": "lammps",
    ".dump": "lammps-dump-text",
    ".gjf": "gaussian-in",
    ".com": "gaussian-in",
    ".fdf": "fdf",
    ".pwi": "espresso-in",
}

TEXT_EXTENSIONS = {
    ".txt",
    ".log",
    ".out",
    ".in",
    ".inp",
    ".sh",
    ".py",
    ".json",
    ".yaml",
    ".yml",
    ".md",
    ".dat",
    ".csv",
    ".err",
    ".xml",
}

TEXT_NAMES = {
    "INCAR",
    "KPOINTS",
    "OSZICAR",
    "POTCAR",
}

ASE_STRUCTURE_NAMES: dict[str, Optional[str]] = {
    "vasp.xml": "vasp-xml",
    "vasprun.xml": "vasp-xml",
}

VASP_STRUCTURE_NAME_RE = re.compile(
    r"(^|[^A-Z0-9])(POSCAR|CONTCAR|XDATCAR|OUTCAR)(?:[0-9]+)?($|[^A-Z0-9])"
)


def is_forced_structure_name(path: Path) -> bool:
    return bool(VASP_STRUCTURE_NAME_RE.search(path.name.upper()))


def detect_preview(path: Path, *, is_dir: bool = False) -> tuple[str, Optional[str]]:
    if is_dir:
        return "directory", None

    lower_name = path.name.lower()
    if lower_name in ASE_STRUCTURE_NAMES:
        return "structure", ASE_STRUCTURE_NAMES[lower_name]

    if is_forced_structure_name(path):
        return "structure", None

    suffix = path.suffix.lower()
    if suffix in TEXT_EXTENSIONS or path.name.upper() in TEXT_NAMES:
        return "text", None

    if suffix in STRUCTURE_EXTENSIONS:
        return "structure", STRUCTURE_EXTENSIONS[suffix]
    return "file", None
